archive/imo/imo2019_q2.lean

```diff
@@ -0,0 +1,576 @@
+or not they follow from the other hypotheses. Similar, a reference to `XY` as a line is taken to
+imply that `X` does not equal `Y` and that is included as a hypothesis, and a reference to `XY`
+being parallel to something is considered a reference to it as a line. However, such an implicit
+hypothesis about two points being different is included only once for any given two points (even
+if it follows from more than one reference to a line or an angle), if `X ≠ Y` is included then
+`Y ≠ X` is not included separately, and such hypotheses are not included in the case where there
+is also a reference in the problem to a triangle including those two points, or to strict
+betweenness of three points including those two. If betweenness is stated, it is taken to be
+strict betweenness. However, segments and sides are taken to include their endpoints (unless
+this makes a problem false), although those degenerate cases might not necessarily have been
+considered when the problem was formulated and contestants might not have been expected to deal
+with them.
+
+We follow Solution 1 from the
+[official solutions](https://www.imo2019.uk/wp-content/uploads/2018/07/solutions-r856.pdf).
+Letting the rays `AA₁` and `BB₁` intersect the circumcircle of `ABC` at `A₂` and `B₂`
+respectively, we show with an angle chase that `P`, `Q`, `A₂`, `B₂` are concyclic and let `ω` be
+the circle through those points. We then show that `C`, `Q₁`, `A₂`, `A₁` are concyclic, and
+then that `Q₁` lies on `ω`, and similarly that `P₁` lies on `ω`, so the required four points are
+concyclic.
+
+-/
+
+noncomputable theory
+
+open affine affine.simplex euclidean_geometry finite_dimensional
+open_locale affine euclidean_geometry real
+
+local attribute [instance] fact_finite_dimensional_of_finrank_eq_succ
+
+variables (V : Type*) (Pt : Type*) [inner_product_space ℝ V] [metric_space Pt]
+variables [normed_add_torsor V Pt] [hd2 : fact (finrank ℝ V = 2)]
+include hd2
+
+/-- A configuration satisfying the conditions of the problem. We define this structure to avoid
+passing many hypotheses around as we build up information about the configuration; the final
+result for a statement of the problem not using this structure is then deduced from one in terms
+of this structure. -/
+@[nolint has_nonempty_instance]
+structure imo2019q2_cfg :=
+(A B C A₁ B₁ P Q P₁ Q₁ : Pt)
+(affine_independent_ABC : affine_independent ℝ ![A, B, C])
+(A₁_mem_BC : A₁ ∈ affine_segment ℝ B C)
+(B₁_mem_AC : B₁ ∈ affine_segment ℝ A C)
+(P_mem_AA₁ : P ∈ affine_segment ℝ A A₁)
+(Q_mem_BB₁ : Q ∈ affine_segment ℝ B B₁)
+(PQ_parallel_AB : line[ℝ, P, Q] ∥ line[ℝ, A, B])
+-- A hypothesis implicit in the named line.
+(P_ne_Q : P ≠ Q)
+(sbtw_P_B₁_P₁ : sbtw ℝ P B₁ P₁)
+(angle_PP₁C_eq_angle_BAC : ∠ P P₁ C = ∠ B A C)
+-- A hypothesis implicit in the first named angle.
+(C_ne_P₁ : C ≠ P₁)
+(sbtw_Q_A₁_Q₁ : sbtw ℝ Q A₁ Q₁)
+(angle_CQ₁Q_eq_angle_CBA : ∠ C Q₁ Q = ∠ C B A)
```

Comment:
Why do you make life harder for yourself by not using `Q Q₁ C` and `A B C` here?? Just because the C's are "inconsistently ordered" in the original equation doesn't mean you have to copy this for your configuration structure, you can just use the ordering of the question at the very end.

Reply:
The angles are consistently ordered in the original problem statement, where consistent means that, although the statement uses unoriented angles, all angles mentioned in the statement have the same orientation in the diagram used in the referenced solution (this is the normal convention for IMO problem statements nowadays). Reversing some of the angles without reversing all of them would lose that consistency, and I don't think avoiding a few uses of `angle_comm` is a significant gain.

Reply:
Aah I see!